### PR TITLE
IMPR/GRID-201

### PR DESCRIPTION
**Pull request by Dwaynekj**

### CODE REVIEW

File: src/dataModels/JSON.js [hunk 1]
```diff
@@ -579,37 +548,9 @@ var JSON = DataModel.extend('dataModels.JSON', {
...
-    },
-
-    /**
-     * @memberOf dataModels.JSON.prototype
      * @param index
      * @param returnAsString
+     * @desc Provides the unicode character used to denote visually if a column is a sorted state
      * @returns {*}
      */
     getSortImageForColumn: function(columnIndex) {
```
Comment (joneit) [inline src/dataModels/JSON.js]: So long as we're doing this, let's do it: Let's move this out too. Put an if around usage (in `getHeaderRowValue`).

File: src/dataModels/JSON.js [hunk 2]
```diff
@@ -537,39 +536,9 @@ var JSON = DataModel.extend('dataModels.JSON', {
...
-                this.applyAnalytics({columnSort: true});
-            }
-        }
-    },
-
     /**
      * @memberOf dataModels.JSON.prototype
+     * @desc returns the columns that currently sorted and their intended direction of the sort
      */
     getSortedColumnIndexes: function() {
```
Comment (joneit) [inline src/dataModels/JSON.js]: So long as we're doing this, let's do it: Let's move this (and behavior version) out too. It won't be called by `applyAnalytics` if there's no sorting datasource.

File: src/Hypergrid.js [hunk 3]
```diff
@@ -1550,6 +1550,20 @@ Hypergrid.prototype = {
         });
         this.canvas.dispatchEvent(clickEvent);
     },
+    /**
+     * @memberOf Hypergrid.prototype
+     * @param {number} c - grid column index.
+     * @param {string[]} keys
+     */
+    fireSyntheticColumnSortEvent: function(c, keys) {
```
Comment (joneit) [inline src/Hypergrid.js]: Why not in mixin?

File: add-ons/hyper-sorter/index.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    });
+
+    function mixInTo(name, instance, mixin) {
+        var object = objects[name];
+        var prototype = object && object.prototype || Object.getPrototypeOf(instance);
+        _(prototype).extend(mixin);
+    }
+}
+
+Hypersorter.prototype = {
```
Comment (joneit) [inline add-ons/hyper-sorter/index.js]: remove this

File: add-ons/hyper-sorter/index.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+'use strict';
+
+var _ = require('object-iterators');
+
+function Hypersorter(grid, objects) {
+    this.grid = grid;
```
Comment (joneit) [inline add-ons/hyper-sorter/index.js]: make this a simple function. doesn't need to be an object with no methods or props:
```javascript
module.exports = function(grid, objects) {
```

File: add-ons/hyper-sorter/index.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    objects = objects || {};
+
+    mixInTo('Hypergrid', grid, require('./mix-ins/grid'));
+    mixInTo('Behavior', grid.behavior, require('./mix-ins/behavior'));
+    mixInTo('Column', grid.behavior, require('./mix-ins/column'));
```
Comment (joneit) [inline add-ons/hyper-sorter/index.js]: this won't work. following will work so long as setData has been called and there is at least one column:
```javascript
mixInTo('Column', grid.behavior.allColumns[0], require('./mix-ins/column'));
```
alternatively, if we have access to the `Column` constructor, perhaps by adding to `src/behaviors/index.js`:
```javascript
objects.Column = Object.getPrototypeOf(grid).constructor.behaviors.Column;
mixInTo('Column', undefined, require('./mix-ins/column'));
```

File: src/Hypergrid.js (hunk 3 above)
Comment (Dwaynekj) [inline src/Hypergrid.js]: We will need the grid to trigger the events. The application developer will listen to these events and decide to act (or not) and if acting means doing a setData or applyAnaltyics

File: add-ons/hyper-sorter/index.js (hunk 6 above)
Comment (Dwaynekj) [inline add-ons/hyper-sorter/index.js]: Sorry I wasn't in the middle of changing this. I should held off on the PR a little longer

File: add-ons/hyper-sorter/index.js (hunk 4 above)
Comment (Dwaynekj) [inline add-ons/hyper-sorter/index.js]: ok

File: src/dataModels/JSON.js (hunk 1 above)
Comment (Dwaynekj) [inline src/dataModels/JSON.js]: In terms of Filtering UI and Sort UI, your saying leave that up to the user?

File: src/dataModels/JSON.js (hunk 2 above)
Comment (Dwaynekj) [inline src/dataModels/JSON.js]: this is the "state" that would be used for serialization

File: src/dataModels/JSON.js (hunk 1 above)
Comment (joneit) [inline src/dataModels/JSON.js]: All I was saying was if we're doing this mixin thing, let's not leave any random "sorting" methods in the core.

File: src/dataModels/JSON.js (hunk 2 above)
Comment (joneit) [inline src/dataModels/JSON.js]: yup just saying let's move it out

File: src/Hypergrid.js (hunk 3 above)
Comment (joneit) [inline src/Hypergrid.js]: this is fine, although the present task was just to moving sorting methods out, and in this case you have added one, so I'm just saying add it to the mix-in instead.

File: src/dataModels/JSON.js (hunk 2 above)
Comment (joneit) [inline src/dataModels/JSON.js]: yup is there some reason in cannot or should not be in the mixin file?

File: src/dataModels/JSON.js (hunk 1 above)
Comment (joneit) [inline src/dataModels/JSON.js]: no just saying move it out

File: src/dataModels/JSON.js (hunk 2 above)
Comment (Dwaynekj) [inline src/dataModels/JSON.js]: I was just thinking something like filter and sort state were agnostic or the dataSource/app logic -> setData called that handles it. So they wouldn't be apart of any mixin but the grid table state?

I wonder what filter state would look like? Maybe a string of the characters supplied for filtering for all columns

Comment (Dwaynekj) [inline src/dataModels/JSON.js]: I thing we would need also a getFilteredColumnState. Do you understand what I mean as it being just agnostic grid UI state

File: src/dataModels/JSON.js (hunk 1 above)
Comment (Dwaynekj) [inline src/dataModels/JSON.js]: ok. That makes sense. If a user wants to customize the column headers to show column state they can certainly do that logic

File: src/Hypergrid.js (hunk 3 above)
Comment (Dwaynekj) [inline src/Hypergrid.js]: We need an API for the user to emit a sort and filtering events (fireSyntheticColumnFilterEvent or the like) as much as we need a cleaner event model. How they user chooses to respond to them was the point of the mixins, I thought.

Maybe you want to go with a model, where the user chooses their own event namespaces (outside of core UI events) and building their own model?

File: src/dataModels/JSON.js (hunk 2 above)
Comment (Dwaynekj) [inline src/dataModels/JSON.js]: Then again. Maybe we can just make the tableState object very generic and the user can use it as a central storage area for their custom application properties and we don't have any sort/filter state. We could make an area called `tablestate.customs = {}`. Thoughts?

File: add-ons/hyper-sorter/index.js (hunk 5 above)
Comment (Dwaynekj) [inline add-ons/hyper-sorter/index.js]: Done. Although I still `new` it in the pink demo

**Comment by Dwaynekj** [comment]
Will merge, Since your PR is based on it https://github.com/openfin/fin-hypergrid/pull/336

**Comment by joneit** [comment]
no, #336 is _not_ based on this...

**Pull request merged**